File: test/SaltarelleProjectTests/CoreSubset/Core/Utility/CssDictionary.cs
// ---------------------------------------------------------------------------------------------------------------------
// <copyright file="CssDictionary.cs" company="Tableau Software">
//   This file is the copyrighted property of Tableau Software and is protected by registered patents and other
//   applicable U.S. and international laws and regulations.
//
//   Unlicensed use of the contents of this file is prohibited. Please refer to the NOTICES.txt file for further details.
// </copyright>
// ---------------------------------------------------------------------------------------------------------------------

namespace Tableau.JavaScript.Vql.Core
{
    using System;
    using System.Collections;
    using System.Collections.Generic;
    using System.Diagnostics.CodeAnalysis;
    using System.Runtime.CompilerServices;

    using jQueryApi;

    [Imported, NamedValues]
    public enum CssPosition
    {
        Static,
        Relative,
        Absolute,
        Fixed,
        Inherit
    }

    [Imported, NamedValues]
    public enum CssSizeKeyword
    {
        [ScriptName("border-box")]
        BorderBox,
        [ScriptName("content-box")]
        ContentBox,
        Available,
        [ScriptName("min-content")]
        MinContent,
        [ScriptName("max-content")]
        MaxContent,
        [ScriptName("fit-content")]
        FitContent,
        Complex,
        Auto
    }

    [Imported, NamedValues]
    public enum CSSTextOverflow
    {
        Clip,
        Ellipsis,
        Inherit,
        Initial
    }

    [Imported, NamedValues]
    public enum CssOverflow
    {
        Visible,
        Hidden,
        Scroll,
        Auto
    }

    [Imported]
    [NamedValues]
    public enum CssPointerEvent
    {
        Auto,
        None,
        VisiblePainted,
        VisibleFill,
        VisibleStroke,
        Visible,
        Painted,
        Fill,
        Stroke,
        All,
        Inherit
    }

    [Imported, NamedValues]
    public enum CssVisibility
    {
        Visible,
        Hidden,
        Collapse
    }

    [Imported, NamedValues]
    public enum CssDisplay
    {
        None,
        Inline,
        Block,
        Flex,
        [ScriptName("inline-block")]
        InlineBlock,
        [ScriptName("table-cell")]
        TableCell,
        Table
    }

    [Imported, NamedValues]
    public enum CssTextAlign
    {
        Left,
        Right,
        Center,
        Justify
    }

    [Imported, NamedValues]
    public enum CssWhiteSpace
    {
        Normal,
        Nowrap,
        Pre,
        [ScriptName("pre-wrap")]
        PreWrap,
        [ScriptName("pre-line")]
        PreLine
    }

    [Imported, NamedValues]
    public enum CssTextDecorationLine
    {
        None,
        Underline,
        Overline,
        [ScriptName("line-through")]
        LineThrough
    }

    /// <summary>
    /// Utility class for giving us a compiler safe way of specifying CSS attributes to be passed to
    /// <see cref="jQueryObject.CSS(JsDictionary)"/>.  All of this should compile away and at runtime we're just
    /// creating JS objects.
    /// Please add new CSS properties to this class as you need them.
    /// </summary>
    /// <example>
    /// The following code:
    /// <code>
    /// this.Template.DomRoot.CSS(new CssDictionary
    /// {
    ///     Width = CssSize.Px(zoneVm.Width),
    ///     Height = zoneVm.Height.AsPx(),
    ///     Top = zoneVm.Y.AsPx(),
    ///     Left = zoneVm.X.AsPx(),
    /// });
    /// </code>
    /// would compile down to:
    /// <code>
    /// this.get_template().get_domRoot().css({
    ///     width: zoneVm.get_width() + 'px',
    ///     height: zoneVm.get_height() + 'px',
    ///     top: zoneVm.get_y() + 'px',
    ///     left: zoneVm.get_x() + 'px'
    /// });
    /// </code>
    /// </example>
    [Imported, Serializable]
    public class CssDictionary
    {
        public TypeOption<CssPosition, string> Position;

        public TypeOption<CssSize, string> Top;

        public TypeOption<CssSize, string> Left;

        public TypeOption<CssSize, string> Right;

        public TypeOption<CssSize, string> Bottom;

        public TypeOption<CssSize, CssSizeKeyword, string> Width;

        public TypeOption<CssSize, CssSizeKeyword, string> Height;

        public TypeOption<CssSize, string> Margin;

        [ScriptName("margin-top")]
        public TypeOption<CssSize, string> MarginTop;

        [ScriptName("margin-left")]
        public TypeOption<CssSize, string> MarginLeft;

        [ScriptName("margin-bottom")]
        public TypeOption<CssSize, string> MarginBottom;

        [ScriptName("margin-right")]
        public TypeOption<CssSize, string> MarginRight;

        [ScriptName("max-height")]
        public TypeOption<CssSize, string> MaxHeight;

        [ScriptName("z-index")]
        public string ZIndex;

        [ScriptName("background-color")]
        public string BackgroundColor;

        public string Color;

        public string Opacity;

        [ScriptName("font-family")]
        public string FontFamily;

        [ScriptName("font-size")]
        public TypeOption<CssSize, string> FontSize;

        [ScriptName("font-style")]
        public string FontStyle;

        [ScriptName("font-weight")]
        public string FontWeight;

        [ScriptName("text-decoration")]
        public TypeOption<CssTextDecorationLine, string> TextDecoration;

        [ScriptName("text-decoration-line")]
        public TypeOption<CssTextDecorationLine, string> TextDecorationLine;

        [ScriptName("white-space")]
        public TypeOption<CssWhiteSpace, string> WhiteSpace;

        // PhantomJS converts padding to padding-top, padding-right and etc
        // use explicit padding properties to avoid unit test surprise failures
        [ScriptName("padding-top")]
        public TypeOption<CssSize, string> PaddingTop;

        [ScriptName("padding-right")]
        public TypeOption<CssSize, string> PaddingRight;

        [ScriptName("padding-bottom")]
        public TypeOption<CssSize, string> PaddingBottom;

        [ScriptName("padding-left")]
        public TypeOption<CssSize, string> PaddingLeft;

        public string Border;

        [ScriptName("box-shadow")]
        public string BoxShadow;

        [ScriptName("border-width")]
        public TypeOption<CssSize, string> BorderWidth;

        [ScriptName("border-top-width")]
        public TypeOption<CssSize, string> BorderTopWidth;

        [ScriptName("border-left-width")]
        public TypeOption<CssSize, string> BorderLeftWidth;

        [ScriptName("border-bottom-width")]
        public TypeOption<CssSize, string> BorderBottomWidth;

        [ScriptName("border-right-width")]
        public TypeOption<CssSize, string> BorderRightWidth;

        [ScriptName("border-color")]
        public string BorderColor;

        [ScriptName("border-top-color")]
        public string BorderTopColor;

        [ScriptName("border-left-color")]
        public string BorderLeftColor;

        [ScriptName("border-bottom-color")]
        public string BorderBottomColor;

        [ScriptName("border-right-color")]
        public string BorderRightColor;

        [ScriptName("border-style")]
        public string BorderStyle;

        [ScriptName("border-top-style")]
        public string BorderTopStyle;

        [ScriptName("border-left-style")]
        public string BorderLeftStyle;

        [ScriptName("border-bottom-style")]
        public string BorderBottomStyle;

        [ScriptName("border-right-style")]
        public string BorderRightStyle;

        [ScriptName("text-align")]
        public TypeOption<CssTextAlign, string> TextAlign;

        [ScriptName("line-height")]
        public string LineHeight;

        [ScriptName("box-sizing")]
        public CssSizeKeyword BoxSizing;

        public TypeOption<CssDisplay, string> Display;

        [ScriptName("pointer-events")]
        public TypeOption<CssPointerEvent, string> PointerEvents;

        public TypeOption<CssVisibility, string> Visibility;

        public TypeOption<CSSTextOverflow, string> TextOverflow;

        public TypeOption<CssOverflow, string> Overflow;

        [ScriptName("overflow-x")]
        public TypeOption<CssOverflow, string> OverflowX;

        [ScriptName("overflow-y")]
        public TypeOption<CssOverflow, string> OverflowY;

        [ScriptName("flex-grow")]
        public string FlexGrow;

        [ScriptName("flex-shrink")]
        public string FlexShrink;

        [ScriptName("flex-basis")]
        public TypeOption<CssSize, string> FlexBasis;

        [ScriptName("order")]
        public string Order;

        [ScriptName("transform")]
        public string Transform;

        [ScriptName("transform-origin")]
        public string TransformOrigin;

        [ObjectLiteral]
        public CssDictionary() { }

        [InlineCode("{d}")]
        public static implicit operator JsDictionary(CssDictionary d)
        {
            return d;
        }

        [InlineCode("{d}")]
        public static implicit operator JsDictionary<string, string>(CssDictionary d)
        {
            return d;
        }
    }

    [Imported, Serializable]
    public sealed class CssSize
    {
        private CssSize() { }

        [ScriptSkip]
        public static implicit operator string(CssSize s)
        {
            return s;
        }
    }

    [Imported]
    public static class CssHelpers
    {
        [InlineCode("{i} + 'px'"), SuppressMessage("Microsoft.Usage", "CA1801:ReviewUnusedParameters", MessageId = "i")]
        public static CssSize AsPx(this int i)
        {
            return null;
        }

        [InlineCode("{i} + 'px'"), SuppressMessage("Microsoft.Usage", "CA1801:ReviewUnusedParameters", MessageId = "i")]
        public static CssSize AsPx(this uint i)
        {
            return null;
        }

        [InlineCode("{i} + 'px'"), SuppressMessage("Microsoft.Usage", "CA1801:ReviewUnusedParameters", MessageId = "i")]
        public static CssSize AsPx(this double i)
        {
            return null;
        }

        [InlineCode("{i} + 'pt'"), SuppressMessage("Microsoft.Usage", "CA1801:ReviewUnusedParameters", MessageId = "i")]
        public static CssSize AsPt(this int i)
        {
            return null;
        }

        [InlineCode("{i} + 'pt'"), SuppressMessage("Microsoft.Usage", "CA1801:ReviewUnusedParameters", MessageId = "i")]
        public static CssSize AsPt(this uint i)
        {
            return null;
        }

        [InlineCode("'scale(' + {x} + ',' + {y} + ')'"), SuppressMessage("Microsoft.Usage", "CA1801", MessageId = "x"), SuppressMessage("Microsoft.Usage", "CA1801", MessageId = "y")]
        public static string AsTransformScale(float x, float y)
        {
            return null;
        }
    }
}

File: E2ETestResults/CoreSubset/Core/Utility/CssDictionary.ts
export const enum CssPosition {
  Static = 'static',
  Relative = 'relative',
  Absolute = 'absolute',
  Fixed = 'fixed',
  Inherit = 'inherit',
}

export const enum CssSizeKeyword {
  BorderBox = 'border-box',
  ContentBox = 'content-box',
  Available = 'available',
  MinContent = 'min-content',
  MaxContent = 'max-content',
  FitContent = 'fit-content',
  Complex = 'complex',
  Auto = 'auto',
}

export const enum CSSTextOverflow {
  Clip = 'clip',
  Ellipsis = 'ellipsis',
  Inherit = 'inherit',
  Initial = 'initial',
}

export const enum CssOverflow {
  Visible = 'visible',
  Hidden = 'hidden',
  Scroll = 'scroll',
  Auto = 'auto',
}

export const enum CssPointerEvent {
  Auto = 'auto',
  None = 'none',
  VisiblePainted = 'visiblePainted',
  VisibleFill = 'visibleFill',
  VisibleStroke = 'visibleStroke',
  Visible = 'visible',
  Painted = 'painted',
  Fill = 'fill',
  Stroke = 'stroke',
  All = 'all',
  Inherit = 'inherit',
}

export const enum CssVisibility {
  Visible = 'visible',
  Hidden = 'hidden',
  Collapse = 'collapse',
}

export const enum CssDisplay {
  None = 'none',
  Inline = 'inline',
  Block = 'block',
  Flex = 'flex',
  InlineBlock = 'inline-block',
  TableCell = 'table-cell',
  Table = 'table',
}

export const enum CssTextAlign {
  Left = 'left',
  Right = 'right',
  Center = 'center',
  Justify = 'justify',
}

export const enum CssWhiteSpace {
  Normal = 'normal',
  Nowrap = 'nowrap',
  Pre = 'pre',
  PreWrap = 'pre-wrap',
  PreLine = 'pre-line',
}

export const enum CssTextDecorationLine {
  None = 'none',
  Underline = 'underline',
  Overline = 'overline',
  LineThrough = 'line-through',
}

/**
 * Utility class for giving us a compiler safe way of specifying CSS attributes to be passed to
 * {@link jQueryObject.CSS}.  All of this should compile away and at runtime we're just
 * creating JS objects.
 * Please add new CSS properties to this class as you need them.
 * @example The following code:
 * `
 * this.Template.DomRoot.CSS(new CssDictionary
 * {
 * Width = CssSize.Px(zoneVm.Width),
 * Height = zoneVm.Height.AsPx(),
 * Top = zoneVm.Y.AsPx(),
 * Left = zoneVm.X.AsPx(),
 * });
 * `
 * would compile down to:
 * `
 * this.get_template().get_domRoot().css({
 * width: zoneVm.get_width() + 'px',
 * height: zoneVm.get_height() + 'px',
 * top: zoneVm.get_y() + 'px',
 * left: zoneVm.get_x() + 'px'
 * });
 * `
 */
export class CssDictionary {
  public position: Object<CssPosition, string>;

  public top: Object<CssSize, string>;

  public left: Object<CssSize, string>;

  public right: Object<CssSize, string>;

  public bottom: Object<CssSize, string>;

  public width: Object<CssSize, CssSizeKeyword, string>;

  public height: Object<CssSize, CssSizeKeyword, string>;

  public margin: Object<CssSize, string>;

  public margin-top: Object<CssSize, string>;

  public margin-left: Object<CssSize, string>;

  public margin-bottom: Object<CssSize, string>;

  public margin-right: Object<CssSize, string>;

  public max-height: Object<CssSize, string>;

  public z-index: string;

  public background-color: string;

  public color: string;

  public opacity: string;

  public font-family: string;

  public font-size: Object<CssSize, string>;

  public font-style: string;

  public font-weight: string;

  public text-decoration: Object<CssTextDecorationLine, string>;

  public text-decoration-line: Object<CssTextDecorationLine, string>;

  public white-space: Object<CssWhiteSpace, string>;

  public padding-top: Object<CssSize, string>;

  public padding-right: Object<CssSize, string>;

  public padding-bottom: Object<CssSize, string>;

  public padding-left: Object<CssSize, string>;

  public border: string;

  public box-shadow: string;

  public border-width: Object<CssSize, string>;

  public border-top-width: Object<CssSize, string>;

  public border-left-width: Object<CssSize, string>;

  public border-bottom-width: Object<CssSize, string>;

  public border-right-width: Object<CssSize, string>;

  public border-color: string;

  public border-top-color: string;

  public border-left-color: string;

  public border-bottom-color: string;

  public border-right-color: string;

  public border-style: string;

  public border-top-style: string;

  public border-left-style: string;

  public border-bottom-style: string;

  public border-right-style: string;

  public text-align: Object<CssTextAlign, string>;

  public line-height: string;

  public box-sizing: CssSizeKeyword;

  public display: Object<CssDisplay, string>;

  public pointer-events: Object<CssPointerEvent, string>;

  public visibility: Object<CssVisibility, string>;

  public textOverflow: Object<CSSTextOverflow, string>;

  public overflow: Object<CssOverflow, string>;

  public overflow-x: Object<CssOverflow, string>;

  public overflow-y: Object<CssOverflow, string>;

  public flex-grow: string;

  public flex-shrink: string;

  public flex-basis: Object<CssSize, string>;

  public order: string;

  public transform: string;

  public transform-origin: string;

  public constructor() { }
}

export class CssSize {
  private constructor() { }

  public static op_Implicit(s: CssSize): string {
    return s;
  }
}

export class CssHelpers {
}
